packages/fast-components-react-base/src/tabs/tabs.tsx

```diff
@@ -0,0 +1,336 @@
+                        onClick: this.handleClick,
+                        onKeyDown: this.handleKeyDown,
+                        tabIndex: this.state.activeId !== tabItem.props.id ? -1 : 0
+                    }
+                )
+            );
+        });
+
+        return tabElements;
+    }
+
+    /**
+     * Renders the tab panels
+     */
+    private renderTabPanels(): JSX.Element[] {
+        const tabPanels: JSX.Element[] = [];
+
+        this.getChildBySlot(
+            this.props.children,
+            this.getSlot(TabSlot.tabItem)
+        ).forEach((tabItem: JSX.Element, index: number): void => {
+            tabPanels.push(
+                React.cloneElement(
+                    this.getChildBySlot(
+                        tabItem.props.children,
+                        this.getSlot(TabSlot.tabPanel)
+                    )[0],
+                    {
+                        key: tabItem.props.id,
+                        id: tabItem.props.id,
+                        "aria-labelledby": tabItem.props.id,
+                        "aria-hidden": this.state.activeId !== tabItem.props.id
+                    }
+                )
+            );
+        });
+
+        return tabPanels;
+    }
+
+    /**
+     * Handles the click event on the tab element
+     */
+    private handleClick = (e: React.MouseEvent<HTMLButtonElement>): void => {
+        if (!this.props.activeId) {
+            this.setState({
+                activeId: e.currentTarget.getAttribute("aria-controls")
+            });
+        } else if (typeof this.props.onUpdateTab === "function") {
+            this.props.onUpdateTab(e.currentTarget.getAttribute("aria-controls"));
+        }
+    }
+
+    /**
+     * Handles the keydown event on the tab element
+     */
+    private handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>): void => {
+        switch (e.keyCode) {
+            case KeyCodes.arrowLeft:
+            case KeyCodes.arrowUp:
+                this.activatePrevious();
+                break;
+            case KeyCodes.arrowRight:
+            case KeyCodes.arrowDown:
+                this.activateNext();
+                break;
+            case KeyCodes.home:
+                this.activateFirst();
+                break;
+            case KeyCodes.end:
+                this.activateLast();
+                break;
+        }
+    }
+
+    /**
+     * Activates the previous tab item
+     */
+    private activatePrevious(): void {
+        const items: JSX.Element[] = this.getChildBySlot(
+            this.props.children,
+            this.getSlot(TabSlot.tabItem)
+        );
+        const currentItemIndex: number = items.findIndex(this.getCurrentIndexById);
+        const previousItemIndex: number = currentItemIndex > 0 ? currentItemIndex - 1 : items.length - 1;
+        const previousItemId: string = items[previousItemIndex].props.id;
+
+        if (!this.props.activeId) {
```

Comment:
All of the "activate" methods share a common pattern of updating state to a particular index and focusing the element. I would reccomend centralizing that logic into a discrete function to reduce duplication and complexity